### PR TITLE
Merge master into release-1.3

### PR DESCRIPTION
**Pull request by rshriram**
Due to lots of code moves in master, its hard to pick out specific
PRs that need to be cherrypicked. So merging master as is. We have
not had any major features land in master since the creation of release-1.3 branch.

Changes:

7e1b986d3 (Johannes Scheuermann, 6 hours ago)
   Update istiocoredns version to a recent release (#16417)

7c4e97c8b (Shriram Rajagopalan, 9 hours ago)
   EnvoyFilter: bug fix for merge ops on ANY type fields (#16419)

   Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>
18da209f2 (leo, 12 hours ago)
   correct some minor nitpicks (#16192)

56c06b11e (Zhonghu Xu, 13 hours ago)
   Prevent duplicate envoy filters when proxy ns equals root config ns
   (#16403)

   * Prevent duplicate envoy filters when proxy ns equals root config ns

   * fix lint

0104d9d57 (Sam Naser, 14 hours ago)
   Remove misplaced service.go file description from beneath various licenses
   (#16420)

ad61658d5 (Yuchen Dai, 16 hours ago)
   pilot: capture all ports by default (#16414)

   Signed-off-by: Yuchen Dai <silentdai@gmail.com>
ea82718ce (Johannes Scheuermann, 17 hours ago)
   Fix line ending for yaml files (#16415)

fd80f5d11 (Sam Naser, 20 hours ago)
   Add supporting implementation for ISTIO_MUTUAL Gateway TLS mode (#15941)

   * Add supporting implementation for ISTIO_MUTUAL Gateway TLS mode

   * Fix erroneous usage of client certs with ISTIO_MUTUAL, use server certs 
   instead

   * Cover ISTIO_MUTUAL validation logic in test case

c787ef27d (Shriram Rajagopalan, 20 hours ago)
   Bug fixes in EnvoyFilter (#16395)

   * Bug fixes in EnvoyFilter

   Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>

   * lint

   Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>

   * integration test

b6cb14e8b (Jonh Wendell, 22 hours ago)
   Upgrade helm to the recommended minimum version: 2.10 (#16104)

f269f2d05 (Martin Taillefer, 23 hours ago)
   Update container image. (#16365)

7d40bd408 (John Howard, 25 hours ago)
   Fix typo in injector template trustDomain/meshId (#16385)

5c56816d1 (Yadong Zhang, 34 hours ago)
   Fix err check for service ingress hostname lookup (#16378)

737dcadeb (Pengyuan Bian, 2 days ago)
   Add instruction about stackdriver adapter config (#16388)

d9cd0c079 (Zvi Cahana, 2 days ago)
   Fix data race accessing Proxy.ServiceInstances (#16375)

   * Fix data race accessing Proxy.ServiceInstances

   * Add an error check

454fe7475 (carolynhu, 2 days ago)
   Remove the outdated monitoring package in pilot, replacing it with the
   istio.io/pkg/monitoring package (#16360)

   * Update go modules

   * Update vendor file

   * Remove outdated monitoring package in pilot

   * Update monitoring package to refence istio.io/pkg/monitoring pacakge

6f8b283ec (John Howard, 2 days ago)
   Enable locality lb tests in presubmit (#16380)

abd80871e (Sam Naser, 2 days ago)
   Replace hardcoded cert paths with designated constants (#16002)

3b30a9344 (Sam Naser, 2 days ago)
   Make nodeagent debugendpoint dump certificate data in base64 (#16347)

   For comparison with configuration dump from Envoy, it makes more sense to
   have the debug endpoint dump the base64 encoded x509 certs. In addition,
   add a guard to ensure no null ptr deref in DebugInfo method
ceba93605 (John Howard, 2 days ago)
   Enable race test in integration presubmit (#16368)

   Currently we have a postsubmit job to run the race detector. To me, it 
   doesn't make much sense to have it only as postsubmit, because then we miss
   data races in PRs and have to fix them up later. The cost of adding race
   detector is pretty low (I think -- will measure in this PR), so it should
   be worthwhile.
3ad0fbc78 (Neeraj Poddar, 2 days ago)
   Update proxy SHA to 08d535 (#16354)

   SHA: 08d53544b4465545e095b1446ef8cc58471115e6
a329eea24 (Zhonghu Xu, 2 days ago)
   rm unused `SIDECAR_UID` metadata key (#16371)

7fcc5ce5b (Zhonghu Xu, 2 days ago)
   Fix sidecar takes too long getting initial config (#16338)

   * Fix sidecar takes too long getting initial config

   * no need connection.modelNode.SetServiceInstances here

   * fix dead lock

e9f40d581 (John Howard, 2 days ago)
   Fix SetServiceInstances data race (#16367)

   A recent PR introduced a new call to SetServiceInstances that will 
   concurrently with another call. Since this isn't protected by a mutex
   (and I think that is a good thing), this causes a data race.

   In pushConnection we already call SetServiceInstances, so we don't need to
   set them here as well, and removing this fixes the race.
3c588f9ac (Zvi Cahana, 3 days ago)
   Retrigger ADS push when inbound listeners change (#15950)

   * Fix Controller.AppendInstanceHandler() to actually invoke the handler

   * Trigger ADS push whenever proxy's inbound listeners change

   * Add an error check and log message

   * Addressed review comments

da84a1487 (lei-tang, 4 days ago)
   Prototype of Istio Webhook Controller (#15604)

   * Prototype of Istio Webhook Controller

   This is a prototype of Istio Webhook Controller, which:
   - Sign webhook certificates at k8s CA
   - Manage the certificate lifecycle for a demo webhook.
   - Manage the WebhookConfiguration for a demo webhook.

   * Revise based on review comments

   * Split the code in controller.go

   * Remove service accounts

   * Use configmap to manage the webhookconfiguration

   * Revisions based on reviewer comments

   * Programmatically distinguish webhookconfigurations

   * Add tests for utils.go

   * Add tests to controller.go

   * Add tests to main.go

   * Add a TODO

387dab547 (John Howard, 5 days ago)
   Use standard meshconfig defaults in galley (#16270)

   * Use standard meshconfig defaults

   For some reason galley is using its own special MeshConfig, that is 
   different than Helm/what pilot uses, leading to inconsistent config. This
   changes it to use the standard defaults in pkg/config/mesh.

   * Update golden files

   * Use in meshcfg/defaults

7447c018a (Ozben Evren, 5 days ago)
   Add missing permission to read resources. (#16352)

ca4bb11bb (Frank Budinsky, 5 days ago)
   Fix broken link to docs (#16351)

   * Fix broken link to docs

   * update generated files

1154a7467 (Jimmy Chen, 5 days ago)
   Add test for istio.io task: Secure Ingress Gateway with SDS (#16154)

   * test

   * automate ingress SDS task

   * add scripts

   * revise

   * lint

   * revise

   * revise

   * revise

   * move tests to istio/tests/istio.io/

   * reorganize test structure

   * resolve conflict

   * revise

2a66836a2 (Douglas Reid, 5 days ago)
   feat(node metadata): add bits to extract metadata via sidecar injector
   (#15551)

   * feat(node metadata): add bits to extract metadata via sidecar injector
   
   * add deployment case and update golden files
   
   * rebase cleanup
   
   * wip: update of logic around metadata and owner
   
   * update metadata capture and add exchange keys metadata
   
   * update golden files

c53cd52b4 (Shriram Rajagopalan, 5 days ago)
   Fix UDS and Sidecar ingress listeners issues (#16298)

   * Fix UDS and Sidecar ingress listeners issues
    Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>
   
   * lint
    Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>
   
   * fixes
    Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>
   
   * lint
    Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>
   
   * listener fixes
   
   * tests
    Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>

69bfc2cf2 (John Howard, 5 days ago)
   Skip mixer native race tests (#16343)

   * Skip mixer native race tests

   * Require kube instead

   * Remove race test env

549b0aa84 (Nino Kodabande, 5 days ago)
   Remove random git merge indicators (#16344)

38e023c41 (Zhonghu Xu, 5 days ago)
   Services without selector incremental eds fix (#16339)

09ba62a10 (John Howard, 5 days ago)
   Fix TestNamespaceTargeting testing.T passing (#16323)

   Prior to this, the parents testing.T was passed, so when a test failed it
   would indicate the parent failed, but not which child, which was confusing.
45a3140ac (John Howard, 6 days ago)
   Move kube/inject out of Pilot (#16284)

   * Move pilot/pkg/kube/inject -> pkg/kube/inject

   * Break out sidecar-injector to top level

6f4be1d55 (Ozben Evren, 6 days ago)
   Update gRPC dependency. (#16322)

   * Update gRPC dependency.

   * Update api package.

   * Update api package.

   * Suppress deprecation warning.

   * Fix mixer integration test.

86d46e720 (Yangmin Zhu, 6 days ago)
   add AuthorizationPolicy CRD (#16187)

5a3ed2b9a (Ed Snible, 6 days ago)
   Replace confusing name with meaningful name (#16319)

3bcfe3fa7 (Nino Kodabande, 6 days ago)
   enable serviceDiscovery in galley's template (#16279)

   * enable serviceDiscovery in galley's template

   * Address review comments

08410f087 (Jimmy Chen, 6 days ago)
   extend wait time for gateway listener to be active. (#16318)

38634af69 (Ed Snible, 6 days ago)
   Upgrade to Werkzeug 0.15 to handle unknown user ids (#15998)

17062f4de (John Howard, 6 days ago)
   Add retries to reachability tests (#16317)

f8a317bb0 (Kuat, 6 days ago)
   mixer: remove HTTPAPISpec (#16315)

   * remove HTTPAPISpec from mixer

   Signed-off-by: Kuat Yessenov <kuat@google.com>

   * lint

   Signed-off-by: Kuat Yessenov <kuat@google.com>

512679316 (Yuchen Dai, 6 days ago)
   sidecar: testcase should not add service entry bind to loopback address
   (#16177)

   * sidecar: testcase should not add service entry bind to loopback address
    Signed-off-by: Yuchen Dai <silentdai@gmail.com>
   
   * tests: use 192.0.2.0/24 addresses in fake service entry
    Signed-off-by: Yuchen Dai <silentdai@gmail.com>

e41dba9d1 (Nathan Mittler, 6 days ago)
   Moving envoy startup code out of pilot (#16300)

   We have a few places in the codebase that duplicate the logic for starting
   up envoy instances. This is the first part of an effort to merge all of
   these into a common utility.  For now, this is just moving the code as-is.
cc712da59 (Nathan Mittler, 6 days ago)
   Autogenerating Istio schemas from yaml. (#16281)

   This follows the pattern in Galley and could potentially help to merge the
   two in the future.
979015470 (Kuat, 6 days ago)
   mixer: update test certificates (#16206)

   * mixer: update test certificates

   Signed-off-by: Kuat Yessenov <kuat@google.com>

   * update to 5 years

   Signed-off-by: Kuat Yessenov <kuat@google.com>

12460093a (Sam Naser, 6 days ago)
   Initial Citadel namespace targeting integration test (#16090)

7e6a56e86 (Zhonghu Xu, 6 days ago)
   Stop using deprecated api field in RedisProxy (#16289)

   * Stop using deprecated redis field
   
   * fix ut

b7f7edb95 (Yan Xue, 6 days ago)
   remove h2 (#16291)

17192488d (Kebe, 6 days ago)
   Add unix-socket support for istioctl pc endpoint (#16288)

   * Add unix-socket support for istioctl pc endpoint

   * Fix golint

332bf64a4 (John Howard, 6 days ago)
   Align egressgateway resource limit with ingress (#16272)

   Having such a low limit leads to OOM on a moderate sized cluster. This PR
   brings it in line with ingressgateway defaults.
cdc0fac96 (yangdihangN, 6 days ago)
   Check ip for ingress/router type sidecar (#16229)

409fb9765 (Bob - Bui Nguyen Thang, 7 days ago)
   add missing priorityClassName for telemetry deployment (#16204)

   The mixer telemetry deployment spec missing priorityClassName taken from
   .Values.global.priorityClassName
09a027ec7 (Kuat, 7 days ago)
   telemetry: add mesh ID (#16116)

   * telemetry: add mesh ID

   Signed-off-by: Kuat Yessenov <kuat@google.com>

   * helm does not understand unicode

   Signed-off-by: Kuat Yessenov <kuat@google.com>

fc89c9b05 (Quanjie Lin, 7 days ago)
   add header supporting google ca regional routing  (#16202)

   * region

   * zone only

   * fix test

6fcfcd5bc (Yuchen Dai, 7 days ago)
   pilot: fix dst ip matching for passthrough clusters (#16255)

   * pilot: fix dst ip matching for passthrough clusters
    Signed-off-by: Yuchen Dai <silentdai@gmail.com>
   
   * fix lint
    Signed-off-by: Yuchen Dai <silentdai@gmail.com>

d51f81dd1 (Nathan Mittler, 7 days ago)
   Moving config schema out of pilot (#16061)

c58f4bc7c (John Howard, 7 days ago)
   Fix new installer test (#16258)

fe43cecc5 (Yan Xue, 8 days ago)
   add protocol detection timeout (#16223)

   * add protocol detection timeout

   * enabled for all listener filters

   * test

   * resolve comments

   * add check

   * add global value

   * fix

   * ..

   * revert change

   * add missing proxy

   * add missing env

1fed9cdd0 (Martin Ostrowski, 8 days ago)
   Add operator/installer artifacts to release tar (#16208)

   * Add operator/installer artifacts to release tar

   * Fix version

4f8e0e431 (Yan Xue, 8 days ago)
   upgrade istio/proxy to 47e4559b8e4f0d516c0d17b233d127a3deb3d7ce (#16248)

4d193dc40 (John Howard, 8 days ago)
   Add legacy varient to kind config patch (#16176)

   * Add legacy varient to kind config patch

   The config patch provided only works on 1.15, but we test on older 
   kubernetes versions for backwards compatibility, so we need another variant
   to support these.

   * Fix comment

1d3fb572f (John Howard, 8 days ago)
   Check proxyNeedsPush after recomputing SidecarScope (#16245)

   * option 1 - mutex

   * Revert "option 1 - mutex"

   This reverts commit 3f97b526f5ea67fbd182bf4a4b73574bf25389bd.

   * Check proxyNeedsPush after recomputing SidecarScope

   Without this change, there is both a data race and a logical race condition
   in proxyNeedsPush.

   * SidecarScope is checked outside of a push, and it is not mutex protected,
   so we can read/write to it at the same time if multiple pushes came in
   (this is not hypothetical, the -race tests catch it).
   * SidecarScope is checked for a push, but it doesn't consider that a push
   could be changing the SidecarScope! We need to do this check after the
   SidecarScope has been updated instead.

   We can fix the first with a mutex, but that is a pretty dirty hack and 
   won't resolve the real race condition.

   Rather than a hacky approach of passing of passing more fields between 
   PushRequest and PushEvent, which are already extremely similar, this change
   just merges them together. The only downside I see to this is that before a
   change has been Debounced, it will not have `start` or
   `push` assigned to it. I think this is acceptable though.

   By merging these structs, we can move the proxyNeedsPush check down to just
   before we trigger an actual push, which is after we set SidecarScope
   appropriately.

   * Fix race condition

b030d535d (Shriram Rajagopalan, 8 days ago)
   Add route name to envoy access logs (#16237)

   Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>
c11ef9cc6 (Jason Young, 8 days ago)
   update to latest istio.io/operator (#16241)

   * update to latest istio.io/operator

   Merge the latest operator CLI changes into istioctl. This includes the 
   latest installer charts and values along with several fixes.

   * update vendor

   * add missing vendor file

8cb23f31d (Jimmy Chen, 8 days ago)
   Support ingress component in integration test framework to get HTTPS
   address when running on minikube (#16114)

   * support ingress to run on minikube

   * revise

   * revise

   * revise

   * lint

4728b94bc (John Howard, 8 days ago)
   Fix memory monitor test deadlock (#16224)

   This had a deadlock cases, where we acquire the lock, then try to send on
   the done channel, but we also try to acquire the lock before recieving on
   the done channel. Easy fix here is to release the lock before sending on
   done channel.
5cece4db2 (Yuchen Dai, 8 days ago)
   fix TestNginx on inbound filter chain (#16213)

### CODE REVIEW

**Comment by rshriram** [comment]
@howardjohn yes I just saw that. I have reverted that commit if you look at the commit sequence.

**Comment by rshriram** [comment]
Capture all ports by default must be present in release-1.3. Otherwise there is no point in doing all that work for infinite loop, filter chains etc. talked to Louis Ryan about it. What we cannot enable is the protocol sniffing by default [though I would like to enable it and have a way to disable if people want. Not the other way around].

File: go.mod [hunk 1]
```diff
@@ -155,10 +155,10 @@ require (
 	gopkg.in/square/go-jose.v2 v2.3.1
 	gopkg.in/stack.v1 v1.7.0 // indirect
 	gopkg.in/yaml.v2 v2.2.2
-	istio.io/api v0.0.0-20190813012407-93fc23ae4a32
+	istio.io/api v0.0.0-20190813194754-2745b7cd4150
```
Comment (rlenglet) [inline go.mod]: This commit is not in the `release-1.3` branch of `istio/api`.

File: go.mod [hunk 2]
```diff
@@ -155,10 +155,10 @@ require (
...
 	istio.io/gogo-genproto v0.0.0-20190731221249-06e20ada0df2
 	istio.io/operator v0.0.0-20190813061052-3b21fa19e418
-	istio.io/pkg v0.0.0-20190731230704-fcbac27d69d5
+	istio.io/pkg v0.0.0-20190813222952-67b4f108fe0e
```
Comment (rlenglet) [inline go.mod]: This commit is not in the `release-1.3` branch of `istio/pkg/`.

**Comment by sdake** [comment]
WOW massive code slush.  I have heart burn over the frollowing PRs: https://github.com/istio/istio/pull/15604
the PRs that @howardjohn mentioned

Also this PR will need further backports so expect it:
https://github.com/istio/istio/pull/16365

My knee-jerk reaction is this is a lot of change and to avoid this merge.  Fortunately, an RC has not been cut.  I guess I am good with it @rshriram if the other @istio/release-managers-1-3 are good with it.  I think we need to be very **selective** about future backports after the RC is cut.  I am only keen to introduce absolute blockers or things that improve the build/test/release process.

**Comment by rlenglet** [comment]
> Fortunately, an RC has not been cut.

We were in the process of doing that right now.
We can't have such big changes coming in anymore.

**Comment by rshriram** [comment]
Are you going to cut an RC with all the bugs ? There are a bunch of PRs for release-1.3 branch that are not merged due to CLA bot. What is the criteria for cutting an RC ? The features that are new since 1.2 are still broken and the PRs that fix those aren't merged yet.  :).

**Comment by rlenglet** [comment]
> There are a bunch of PRs for release-1.3 branch that are not merged due to CLA bot.

Is that true? There is currently only one PR waiting for release manager review across all `release-1.3` branches in all repos.

**Comment by duderino** [comment]
> Are you going to cut an RC with all the bugs ? There are a bunch of PRs for release-1.3 branch that are not merged due to CLA bot. What is the criteria for cutting an RC ? The features that are new since 1.2 are still broken and the PRs that fix those aren't merged yet. :).

it would be great to cherry pick the fixes on a bug by bug basis.   If there is anything severe the people who sent them to master should have cherry picked them into 1.3 already, but if people are not doing that I'd love to see where things fell over.

For what it's worth (in a formal sense it's worth nothing since I'm not a 1.3 release manager), I agree with the 1.3 release managers who are pushing back on this PR

**Comment by kyessenov** [comment]
Commit https://github.com/istio/istio/commit/f8a317bb0083b4833a041aa160b5971985f50b74 says in big letters not to be cherrypicked. What is it doing here?

**Comment by fpesce** [comment]
During the 1.2 release cycle, as a release manager then, I took care of cherry-picking and try to put to a higher standard the CI (waiting for each and every PR to have even non-required integration test passing).

The only large cherry picks that were done on 1.2 were cherry picks from the diverged 1.1 branch (because changes landed to 1.1 independently from master), and those cherry picks were on code changes that happened before freeze date (which happens precisely 4 weeks/a month before planned release date, same than for 1.3). This type of large cherry picks should not happen anymore, we now have labels that allow targeted cherry picking per PR, and a reasonable review and quality insurance on code changes.

Thus, I'd recommend to close this PR, please, this is not reasonable and it comes after a first testing session on 1.3 branch that will be considered totally obsolete after such a large change.

**Comment by duderino** [comment]
Closing this in favor of https://github.com/istio/istio/pull/16346

**Pull request closed without merging**